netcam_http.c

```diff
@@ -334,282 +441,289 @@ int netcam_read_next_header(netcam_context_ptr netcam)
-                     * This message is logged as Information as it would be useful to know
-                     * if your netcam often returns bad HTTP result codes.
-                     */
-                    netcam->connect_keepalive = FALSE;
-                    free((void *)netcam->cnt->conf.netcam_keepalive);
-                    netcam->cnt->conf.netcam_keepalive = strdup("off");
-                    MOTION_LOG(NTC, TYPE_NETCAM, NO_ERRNO
-                        ,_("Removed netcam Keep-Alive flag "
-                        "due to apparent closed HTTP connection."));
-                }
-                return ret;
-            }
-            firstflag = 0;
-            free(header);
-            continue;
-        }
-
-        if (*header == 0)   /* Blank line received */
-            break;
+	int retval = -3;      /* "Unknown err" */
+	int ret;
+	int firstflag = 1;
+	int aliveflag = 0;    /* If we have seen a Keep-Alive header from cam. */
+	int closeflag = 0;    /* If we have seen a Connection: close header from cam. */
+	char *header;
+	char *boundary;
+
+			/* Send the initial command to the camera. */
+	if (send(netcam->sock, netcam->connect_request,
+					 strlen(netcam->connect_request), 0) < 0) {
+		MOTION_LOG(ERR, TYPE_NETCAM, SHOW_ERRNO
+							 ,_("Error sending 'connect' request"));
+		return -1;
+	}
+
+			/*
+			 * We expect to get back an HTTP header from the camera.
+			 * Successive calls to header_get will return each line
+			 * of the header received.  We will continue reading until
+			 * a blank line is received.
+			 *
+			 * As we process the header, we are looking for either of
+			 * header lines Content-type or Content-length.  Content-type
+			 * is used to determine whether the camera is "streaming" or
+			 * "non-streaming", and Content-length will be used to determine
+			 * whether future reads of images will be controlled by the
+			 * length specified before the image, or by a boundary string.
+			 *
+			 * The Content-length will only be present "just before" an
+			 * image is sent (if it is present at all).  That means that, if
+			 * this is a "streaming" camera, it will not be present in the
+			 * "first header", but will occur later (after a boundary-string).
+			 * For a non-streaming camera, however, there is no boundary-string,
+			 * and the first header is, in fact, the only header.  In this case,
+			 * there may be a Content-length.
+			 *
+			 */
+	while (1) {     /* 'Do forever' */
+		ret = header_get(netcam, &header, HG_NONE);
+
+		MOTION_LOG(DBG, TYPE_NETCAM, NO_ERRNO,_("Received first header ('%s')"), header);
+
+		if (ret != HG_OK) {
+			MOTION_LOG(WRN, TYPE_NETCAM, NO_ERRNO
+								 ,_("Error reading first header (%s)"), header);
+			free(header);
+			return -1;
+		}
+				
+		if (firstflag) {
+			if ((ret = http_result_code(header)) != 200) {
+				MOTION_LOG(DBG, TYPE_NETCAM, NO_ERRNO,_("HTTP Result code %d"), ret);
+
+				if(ret==401) //try digest auth
```

Comment:
Shouldn't we rely on the server's responce indicating "WWW-Authenticate: Digest"?